### ISSUE TITLE
Support sandbox network mode in rootless

### ISSUE THREAD

**Issue by sfc-gh-cxie** [google/gvisor#10359]
### Description

I'm not sure if it's WAI, is it possible to do runsc run in rootless with sandbox network mode?

Based on my understanding, sandbox network mode is securer but root mode is less securer, is it contradictory? Thanks

**Comment by milantracy** [comment]
https://github.com/google/gvisor/issues/311 answers your second question.

The --rootless option skips a few security layers that can be setup by root only.

**Comment by sfc-gh-cxie** [comment]
> #311 answers your second question.
> 
> The --rootless option skips a few security layers that can be setup by root only.

Thanks Jing, for some context, I wanna run some arbitrary code inside gVisor, if we run it in root mode and arbitrary code break gVisor, it means it will have the root access, that's why we want to run it in rootless mode.

I wonder what would be the suggested way here, should I use host network mode here(which is less secure from the network perspective), or should I try to work on supporting sandbox mode with rootless.

**Comment by EtiennePerot** [comment]
> if we run it in root mode and arbitrary code break gVisor, it means it will have the root access

gVisor re-executes itself as an unprivileged user before starting the container. Therefore, if the code manages to break gVisor, it will not have root access.

When running with `sudo`, gVisor only uses the root privileges in order to create the namespaces and other layers of security (`seccomp` rules, `chroot`, etc) *prior* to entering the sandboxed mode (changing its own user to unprivileged at that point) and starting the container code. I suggest reading the [architecture docs](https://gvisor.dev/docs/) and [security model page](https://gvisor.dev/docs/architecture_guide/security/) for more info.

**Comment by sfc-gh-cxie** [comment]
> > if we run it in root mode and arbitrary code break gVisor, it means it will have the root access
> 
> gVisor re-executes itself as an unprivileged user before starting the container. Therefore, if the code manages to break gVisor, it will not have root access.
> 
> When running with `sudo`, gVisor only temporarily uses the root privileges in order to create the namespaces and other layers of security (`seccomp` rules, `pivot_root`, etc) _prior_ to entering the sandboxed mode (dropping all its privileges and changing its own user to unprivileged at that point) and starting the container code. I suggest reading the [architecture docs](https://gvisor.dev/docs/) and [security model page](https://gvisor.dev/docs/architecture_guide/security/) for more info.

Thanks Etienne, I'm reading the code but seem gVisor re-execute itself in rootless mode?(https://github.com/snowflakedb/gvisor/blob/main/runsc/specutils/namespace.go#L251). But per what you've said, it seems we should always run gVisor in root mode, with more features and extra secure check?

**Comment by EtiennePerot** [comment]
> I'm reading the code but seem gVisor only re-execute itself in rootless mode?(https://github.com/snowflakedb/gvisor/blob/main/runsc/specutils/namespace.go#L251)

Hm, this link goes to a private repository, perhaps Snowflake's fork of gVisor?

The re-execution I am referring to is [here](https://github.com/google/gvisor/blob/c8da73daaf635e7ad372ed6b49ce68e8ab4010b8/runsc/cmd/boot.go#L311), which [sets the UID here](https://github.com/google/gvisor/blob/c8da73daaf635e7ad372ed6b49ce68e8ab4010b8/runsc/cmd/cmd.go#L98) (and drops all capabilities on the next line).

> it seems we should always run gVisor in root mode, with more features and extra secure check?

Yes. Generally speaking, gVisor is intended to take care of doing whatever is the most secure thing to do given whatever privileges it has.

**Comment by EtiennePerot** [comment]
Ah, I see the link is updated. Yes, the code you've linked to re-runs itself as "root" within a new namespace; this is different from "root" in the initial user namespace, and does not grant the process any additional privilege within the initial user namespace (which is where "root" access would be sensitive).

Moreover, this code executes as part of the `runsc run` command. However, the process that runs the actual container is a subprocess of that: it's the `runsc boot` command, which goes through the codepath I've linked to.

**Comment by sfc-gh-cxie** [comment]
That makes sense, thanks. One more clarification on

> gVisor re-executes itself as an unprivileged user before starting the container. Therefore, if the code manages to break gVisor, it will not have root access.

I'm thinking about the case that attacker would escape the container and get the original root privileges. Or are we saying since we've drop all capabilities, even attacker manages escaping from the container, they are unable to get the root privileges either.

**Comment by EtiennePerot** [comment]
> Or are we saying since we've drop all capabilities, even attacker manages escaping from the container, they are unable to get the root privileges either.

Correct. If the code inside the container manages to get the privileges of the `runsc boot` process it runs in, all it can do is whatever permissions the `runsc boot` process can do. But that process has already dropped all of its privileges before running anything inside the container.

> But should we still have the concern about the file access?

Before running any container code, `runsc boot` also runs in a separate mount namespace, and calls `pivot_root` on itself to remove its own access to any host file that isn't a container root filesystem file or mounted volume.

Additionally, if you use `--directfs=false`, the `runsc boot` process's seccomp rules will forbid use of the `open` and `openat` syscall, i.e. it will not be able to open any host file at all.

**Comment by handpickencounter** [comment]
I don't believe there is actually a good reason to deny usage of the gvisor netstack for --rootless.

The problem lies in gvisor creating and changing network namespaces which is root-only.

If there is a way to disable this particular behavior it would help, the real network namespaces in the host's netns can also be hidden then.

This is someone that was done with a virtualization approach:

https://github.com/containers/libkrun

> virtio-vsock + TSI: A novel technique called Transparent Socket Impersonation which allows the VM to have network connectivity without a virtual interface. This technique supports both outgoing and incoming connections. It's possible for userspace applications running in the VM to transparently connect to endpoints outside the VM and receive connections from the outside to ports listening inside the VM. Requires a custom kernel (like the one bundled in libkrunfw) and it's limited to AF_INET SOCK_DGRAM and SOCK_STREAM sockets.

Should be simpler in the gvisor model.

**Comment by nybidari** [comment]
We are actively working on supporting sandbox network mode in rootless environments.

As @EtiennePerot highlighted, while `runsc boot` drops capabilities before running untrusted workloads, setting up 
`--network=sandbox` currently requires elevated privileges (`CAP_NET_ADMIN` / `CAP_SYS_ADMIN`) to manipulate host network namespaces, bind `AF_PACKET` raw sockets, and strip host link IPs. In environments where the outer runtime itself is unprivileged (rootless Kubernetes pods, restricted container environments, or nested sandboxes), `runsc` cannot perform these host netns operations.

### Approach: Decoupling Netstack via UDS
- **UDS / FD-based Link Endpoint**: Instead of binding `AF_PACKET` in a host netns, Netstack inside the sandbox is attached to a file descriptor / Unix Domain Socket (UDS) link endpoint.
- Packets are exchanged across a UDS with a user-space network daemon or proxy running outside the sandbox.
- Eliminates the need for `setns(CLONE_NEWNET)` and veth pair creation on the host.
- Allows flag or OCI spec configuration (`--network=uds` or passing a socket path/FD via flags or OCI annotations).

Happy to hear your thoughts and feedback!

**Comment by handpickencounter** [comment]
That approach is fine. It's also what passt and gvproxy use.

https://github.com/AsahiLinux/muvm for example lets you provide a passt socket.

Something which may be worth including is a multiplexer/server approach such that a single socket can serve multiple gvisor instances.
